File: Daily/2882-ways-to-express-an-integer-as-sum-of-powers/ways-to-express-an-integer-as-sum-of-powers.py
class Solution:
    def numberOfWays(self, n: int, x: int) -> int:
        MOD = 10 ** 9 + 7
        high = ceil(n ** (1 / x))
        # 遞推
        f = [0] * (n + 1)
        f[0] = 1
        for num in range(1, high + 1):
            power = num ** x
            for cap in range(n, power - 1, -1):
                f[cap] = (f[cap] + f[cap - power]) % MOD
        return f[n]
        '''
        @cache
        def dfs(left, num) -> int:
            if left == 0:
                return 1
            if left < 0 or num < 1:
                return 0
            return (dfs(left - num ** x, num - 1) + dfs(left, num - 1)) % MOD

        ans = dfs(n, high)
        dfs.cache_clear()
        return ans
        '''

File: Daily/2882-ways-to-express-an-integer-as-sum-of-powers/ways-to-express-an-integer-as-sum-of-powers.cpp
class Solution {
    using LL = long long;
    LL MOD = 1e9 + 7;
public:
    int numberOfWays(int n, int x) {
        auto quickMul = [&](LL a, LL b) -> LL {
            LL ans = 1;
            while(b) {
                if(b & 1) ans = (ans * a) % MOD;
                a = (a*a) % MOD;
                b >>= 1;
            }

            return ans;
        };

        vector<vector<LL>> memo(n+1, vector<LL>(n, -1));
        function<LL(int i, int sum)> dfs = [&](int i, int sum) -> LL {
            if(sum == n) return 1;
            if(sum > n) return 0;
            if(i > n) return 0;
            if(memo[i][sum] != -1) return memo[i][sum];
            return memo[i][sum] = (dfs(i+1, sum) + dfs(i+1, sum + quickMul(i,x))) % MOD;
        };

        return dfs(1, 0);
    }
};